allennlp/data/dataloader.py

```diff
@@ -20,7 +20,28 @@ def allennlp_collate(instances: List[Instance]) -> TensorDict:
     return batch.as_tensor_dict(batch.get_padding_lengths())
 
 
-class DataLoader(Registrable, data.DataLoader):
+class DataLoader(Registrable):
+    """
+    A `DataLoader` is responsible for generating batches of instances from a `Dataset`,
+    or another source of data. This is essentially just an abstraction over `torch.utils.data.DataLoader`.
+
+    This class only has one required method, `__iter__()`, that creates an iterable
+    of `TensorDict`s. Additionally, this class comes with a `__len__()` method
+    that just raises a `TypeError` by default. When possible, this should be overriden
+    to return the number of batches that will be generated by the `__iter__()` method.
+    """
+
+    default_implementation = "pytorch_dataloader"
+
+    def __len__(self) -> int:
+        raise TypeError
```

Comment:
Why `TypeError`? Is that what the torch one does?

Reply:
Yea, `TypeError` is raised when you use a lazy dataset, and it's also what's raised when `len()` is called on an object that doesn't implement `__len__`. So we'd get the same result by just not including `__len__` in this base class. I just thought it's better to be explicit so that it's obvious `__len__` should be implemented by subclasses whenever possible.